gate-web/src/main/groovy/com/netflix/spinnaker/gate/interceptors/RequestSheddingInterceptor.java

```diff
@@ -0,0 +1,161 @@
+ * requests will be dropped.
+ *
+ * A percentage chance config key is also available which takes a
+ * value between 0 and 100, representing the percentage chance that
+ * the request will be dropped. If the value is set to "60", there
+ * will be a 60% chance the request will be dropped; this value is
+ * only considered when the enabled key is also set to "true".
+ *
+ * If enabled but an internal error occurs due to configuration, the
+ * interceptor will elect to drop all low priority requests.
+ */
+public class RequestSheddingInterceptor extends HandlerInterceptorAdapter {
+
+  final static String PRIORITY_HEADER = "X-Spinnaker-Priority";
+  final static String LOW_PRIORITY = "low";
+
+  final static String ENABLED_KEY = "requestShedding";
+  final static String CHANCE_KEY = "requestShedding.chance";
+  final static String PATHS_KEY = "requestShedding.paths";
+
+  private final static Logger log = LoggerFactory.getLogger(RequestSheddingInterceptor.class);
+  private final static Splitter splitter = Splitter.on(",").omitEmptyStrings().trimResults();
+  private final static Random random = new Random();
+
+  private final DynamicConfigService configService;
+  private final Registry registry;
+
+  private final Id requestsId;
+
+  public RequestSheddingInterceptor(DynamicConfigService configService, Registry registry) {
+    this.configService = configService;
+    this.registry = registry;
+
+    this.requestsId = registry.createId("requestShedding.requests");
+  }
+
+  @Override
+  public boolean preHandle(HttpServletRequest request,
+                           HttpServletResponse response,
+                           Object handler) throws Exception {
+    if (configService.isEnabled(ENABLED_KEY, false) && isDroppable(request)) {
+      if (shouldDropRequestWithChance()) {
+        log.warn("Dropping low priority request: {}", request.getRequestURI());
+        registry.counter(requestsId.withTag("action", "dropped")).increment();
+
+        response.setDateHeader("Retry-After", Instant.now().plus(Duration.ofMinutes(1)).toEpochMilli());
+        throw new LowPriorityRequestRejected();
+      }
+      log.debug("Dice roll prevented low priority request shedding: {}", request.getRequestURI());
+      registry.counter(requestsId.withTag("action", "allowed")).increment();
+    }
+
+    return super.preHandle(request, response, handler);
+  }
+
+  private boolean isDroppable(HttpServletRequest request) {
+    String uri = request.getRequestURI();
+
+    // We want our error handler to service the errors correctly
+    if ("/error".equals(uri)) {
+      return false;
+    }
+
+    return LOW_PRIORITY.equals(request.getHeader(PRIORITY_HEADER)) &&
+      hasPathMatch(uri);
+  }
+
+  private boolean shouldDropRequestWithChance() {
+    Integer chance = configService.getConfig(Integer.class, CHANCE_KEY, 100);
+    if (chance < 0 || chance > 100) {
+      log.warn("Request shedding drop chance is out of bounds: {}, assuming 100", chance);
```

Comment:
this defaulting to 100 combined with no paths matching everything means that it's really easy to accidentally drop everything, no? No concern there?

Reply:
As mentioned above, the clients are telling us what requests are low priority. Yes, it has the potential to drop all low priority requests, but this was the initial requirements to begin with; I only added the chance & path matching as a bonus.

To start, we'll only be adding these headers with a `low` value to the projects page in deck. Furthermore, if a client wants to request the projects endpoints even with load shedding enabled, the endpoint isn't totally blocked off, as they can just not send the header and this interceptor will do nothing.